### PR TITLE
Inline iter_ticks into _update_ticks, and use that in mplot3d.

### PR DESCRIPTION
**Pull request by anntzer**
iter_ticks is only called in one place and the iterator is immediately
converted into a list, so we may as well inline it and create the list
immediately.  Also, we can immediately assign the tick locations and
labels, instead of having to carry around the ticks, the locations and
the labels as a tuple.

Reuse that implementation in axis3d (... which also gains support for
minor ticks for free).

xref https://github.com/matplotlib/matplotlib/pull/13314#discussion_r252351622.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by WeatherGod** [comment]
Just a little bit of a heads-up. We should probably do a sanity check on the image tests, as I think many of them automatically remove tick labels.

File: doc/api/next_api_changes/2018-02-05-AL.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Changes to the internal tick handling API
+`````````````````````````````````````````
+
+``Axis.iter_ticks`` (which only served as a helper to the private
+``Axis._update_ticks``) is deprecated.
+
+The signature of the (private) ``Axis._uodate_ticks`` has been changed to not
```
Comment (QuLogic) [inline doc/api/next_api_changes/2018-02-05-AL.rst]:
```suggestion
The signature of the (private) ``Axis._update_ticks`` has been changed to not
```

Comment (anntzer) [reply doc/api/next_api_changes/2018-02-05-AL.rst]: fixed, thanks

**Comment by anntzer** [comment]
@WeatherGod At least ticks and tick labels on empty 2D and 3D axes look fine.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1012,6 +1012,7 @@ def _set_artist_props(self, a):
             return
         a.set_figure(self.figure)
 
+    @cbook.deprecated("3.1")
     def iter_ticks(self):
```
Comment (jklymak) [inline lib/matplotlib/axis.py]: This looks great, but I wonder if you want a private `_locate_ticks` and then call it here, and in the `_update_ticks`.  That way the (deprecated) public function doesn't drift from the private implimentation?

Comment (anntzer) [reply lib/matplotlib/axis.py]: But then (hypothetical, but that's what this discussion is about) changes in _locate_ticks would be reflected in iter_ticks, and you'd need to add additional changelog entries for changes in the behavior of the deprecated method, which seems a bit silly.

**Comment by anntzer** [comment]
rebased on top of #13314.

**Comment by anntzer** [comment]
Updated another test that suffered from the same issue as https://github.com/matplotlib/matplotlib/pull/13314#issuecomment-458559313 (minor tick overstrike changes).

File: doc/api/next_api_changes/2018-02-05-AL.rst [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Changes to the internal tick handling API
+`````````````````````````````````````````
+
+``Axis.iter_ticks`` (which only served as a helper to the private
+``Axis._update_ticks``) is deprecated.
+
+The signature of the (private) ``Axis._update_ticks`` has been changed to not
```
Comment (timhoffm) [inline doc/api/next_api_changes/2018-02-05-AL.rst]: `axis_artist.AxisArtist._update_ticks()` also takes a `renderer` argument; and it uses it. Not quite clear if we want to keep the APIs similar in both worlds.

Comment (anntzer) [reply doc/api/next_api_changes/2018-02-05-AL.rst]: They don't even return the same kind of things (one returns a list of ticks, the other returns a bounding box) so I wouldn't worry about that.

Comment (timhoffm) [reply doc/api/next_api_changes/2018-02-05-AL.rst]: In that case, I approve.

**Pull request merged**